src/NUnitTestAdapter/TestConverter.cs

```diff
@@ -127,33 +124,81 @@ public IList<VSTestResult> GetVSTestResults(XmlNode resultNode)
         /// </summary>
         private TestCase MakeTestCaseFromXmlNode(XmlNode testNode)
         {
+            var methodName = testNode.GetAttribute("methodname");
+            var displayName = testNode.GetAttribute("name");
+            var className = testNode.GetAttribute("classname");
+
             var testCase = new TestCase(
                                      testNode.GetAttribute("fullname"),
                                      new Uri(NUnitTestAdapter.ExecutorUri),
                                      _sourceAssembly)
             {
-                DisplayName = testNode.GetAttribute("name"),
+                DisplayName = displayName,
                 CodeFilePath = null,
                 LineNumber = 0
             };
 
             if (_collectSourceInformation && _navigationDataProvider != null)
             {
-                var className = testNode.GetAttribute("classname");
-                var methodName = testNode.GetAttribute("methodname");
                 var navData = _navigationDataProvider.GetNavigationData(className, methodName);
                 if (navData.IsValid)
                 {
                     testCase.CodeFilePath = navData.FilePath;
                     testCase.LineNumber = navData.LineNumber;
+                }               
+            }
+            else
+            {
+                bool isTestNameScenario = IsTestNameScenario(methodName, displayName);
+
+                if (isTestNameScenario)
+                {
+                    // Stash property in test case object
+                    testCase.SetPropertyValue(Constants.TestCaseAdjustedFQNProperty, string.Format("{0}.{1}", className, methodName));
                 }
             }
-            
+
             testCase.AddTraitsFromTestNode(testNode, TraitsCache);
 
             return testCase;
         }
 
+        private bool IsTestNameScenario(string testCaseMethodName, string testCaseDisplayName)
+        {
+            if (string.Compare(testCaseMethodName, testCaseDisplayName) == 0)
+            {
+                return false;
+            }
+            else
+            {
+                // Can be TestName, Generic, or Parameterized test case scenario
+                int index = testCaseDisplayName.IndexOf('<');
+
+                if (index > -1)
+                {
+                    // Generic test method
+                    testCaseDisplayName = testCaseDisplayName.Remove(index).TrimEnd();
+                    return string.Compare(testCaseMethodName, testCaseDisplayName) != 0;
+                }
+                else
+                {
+                    // Parameterized test method
+                    index = testCaseDisplayName.IndexOf('(');
+
+                    if (index > -1)
+                    {
+                        testCaseDisplayName = testCaseDisplayName.Remove(index).TrimEnd();
```

Comment:
I think this doesn't properly allow for the fact that the user can put __anything__ into the display name, including parentheses. What if the name is `(`? Crazy but possible and we should not throw. In addition, if the method has arguments, then one of them may be a string containing a left parenthesis.

The code can be fixed to allow for that, of course, returning false if there is any doubt. However, this illustrates for me how fragile it is to try to reverse engineer the NUnit test names. In an early version of the adapter, we did that ourselves and it didn't work for us. That's why we introduced attributes for method and class name. Putting code in the adapter that tries to reverse engineer nunit is a step backward and I think we should try to avoid it by simply using the method and class names provided in all cases.

Reply:
Thanks for the feedback Charlie. Let me see if I can find a better solution. I will get back soon.